### PR TITLE
feat(auth): add SRP sign-in API 

### PR DESCRIPTION
**Pull request by israx**
#### Description of changes

This PR contains a high level `signIn` API that delegates to the granular sign-in implementations based on the auth flow parameter passed in the request (SRP, CUSTOM, etc)

Additionally, this PR includes the sign-in SRP implementation which is delegated by default by the `signIn` API

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/providers/cognito/index.ts [hunk 1]
```diff
@@ -4,11 +4,15 @@
 export { signUp } from './apis/signUp';
 export { resetPassword } from './apis/resetPassword';
 export { confirmResetPassword } from './apis/confirmResetPassword';
+export { signIn } from './apis/signIn';
+export { signInWithSRP } from './apis/signInWithSRP';
```
Comment (elorzafe) [inline packages/auth/src/providers/cognito/index.ts:8]: We don't have to export this on public facing APIs/

File: packages/auth/src/providers/cognito/utils/internalAuthSession.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+// TODO: replace this implementation with state machines
+let internalAuthSession: string | undefined = undefined;
+
+/**
+ * Sets the current active Session used for Cognito APIs during the sign-in process.
+ * @internal
+ */
+export function setInternalAuthSession(session: string | undefined) {
```
Comment (AllanZhengYP) [inline packages/auth/src/providers/cognito/utils/internalAuthSession.ts:11]: According to the purpose, will it be better to be named `setActiveSignInSession()`? `Internal` is a bit over-generalized.

File: packages/auth/src/providers/cognito/utils/signInHelpers.ts [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+export async function handleUserSRPAuthFlow(
+	username: string,
+	password: string,
+	clientMetadata: ClientMetadata | undefined
+): Promise<RespondToAuthChallengeCommandOutput> {
+	const config = Amplify.config;
+	const userPoolId = config['aws_user_pools_id'];
+	const userPoolName = userPoolId.split('_')[1];
+	const authenticationHelper = new AuthenticationHelper(userPoolName);
```
Comment (AllanZhengYP) [inline packages/auth/src/providers/cognito/utils/signInHelpers.ts:41]: This is not a blocking comment -- I hope we can make this class functional in the future.

**Review by AllanZhengYP — APPROVED** [review]
Looks good to me. Only have a minor naming question

**Review by AllanZhengYP — DISMISSED** [review]

File: packages/auth/src/providers/cognito/index.ts [hunk 4]
```diff
@@ -4,11 +4,14 @@
 export { signUp } from './apis/signUp';
 export { resetPassword } from './apis/resetPassword';
 export { confirmResetPassword } from './apis/confirmResetPassword';
+export { signIn } from './apis/signIn';
 
 export { ClientMetadata } from './types/models/ClientMetadata';
 export { CognitoUserAttributeKey } from './types/models/CognitoUserAttributeKey';
 export { CustomAttribute } from './types/models/CustomAttribute';
 export { ValidationData } from './types/models/ValidationData';
+export { AuthFlowType } from './types/models/AuthFlowType';
```
Comment (elorzafe) [inline packages/auth/src/providers/cognito/index.ts:13]: do we need to export all these types?

**Review by elorzafe — CHANGES_REQUESTED** [review]
PR looking good, thanks @israx my only concern is the types that we are exporting will bloat the list of possible things for customers to import. Can we remove those exported types and then add those incrementally when is needed.

File: packages/auth/src/providers/cognito/index.ts (hunk 4 above)
Comment (haverchuck) [reply packages/auth/src/providers/cognito/index.ts:13]: Why don't we just export `AuthFlowType` for now, and we can figure out the rest later?

File: packages/auth/__tests__/providers/cognito/signInWithSRP.test.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+describe('signIn API happy path cases', () => {
+	let handleUserSRPAuthflowSpy;
+
+	beforeEach(() => {
+		handleUserSRPAuthflowSpy = jest
+			.spyOn(initiateAuthHelpers, 'handleUserSRPAuthFlow')
+			.mockImplementationOnce(
+				async (): Promise<RespondToAuthChallengeCommandOutput> => {
+					return authAPITestParams.RespondToAuthChallengeCommandOutput;
+				}
```
Comment (AllanZhengYP) [inline packages/auth/__tests__/providers/cognito/signInWithSRP.test.ts:29]: Very nit:
```suggestion
				async (): Promise<RespondToAuthChallengeCommandOutput> =>
					authAPITestParams.RespondToAuthChallengeCommandOutput;
```

File: packages/auth/src/providers/cognito/types/options/CognitoResetPasswordOptions.ts [hunk 6]
```diff
@@ -1,7 +1,7 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
-import { ClientMetadata } from '../..';
+import { ClientMetadata } from "../models/ClientMetadata";
```
Comment (AllanZhengYP) [inline packages/auth/src/providers/cognito/types/options/CognitoResetPasswordOptions.ts:4]: Very nit: 
```suggestion
import { ClientMetadata } from '../models/ClientMetadata';
```

**Review by AllanZhengYP — APPROVED** [review]

**Review by elorzafe — APPROVED** [review]

**Review by AllanZhengYP — APPROVED** [review]

**Pull request merged**